crossplane/crossplane
# Crossplane to manage its own CRDs

### Description of your changes

This PR adds initialization functionality to core Crossplane process. The steps are:
* Install (or update) all Crossplane CRDs.
* Create `Lock` object.
* Create all `Provider` and `Configuration` packages that are given for preinstallation.

While core Crossplane is doing all these things, RBAC manager has to wait for a few seconds until `CompositeResourceDefinition` and `ProviderRevision` CRDs exist. To do that, it uses an initializer that has only one step that waits for these CRDs.

This will allow Crossplane not depend on Helm to handle CRDs especially during the update. Otherwise, we'd require users to run a `kubectl apply`. Additionally, this initializer mechanism will be useful for generating the TLS Cert for webhook servers.

Fixes https://github.com/crossplane/crossplane/issues/2156


Fixes #

I have:


### How has this code been tested

* Normal installation.
* Install v1.0.0, install provider-aws, provision a `VPC`. Then upgrade using `cluster/local/kind.sh helm-upgrade`.
* Install v1.0.0, follow quick start [composition](https://crossplane.io/docs/v1.0/getting-started/compose-infrastructure.html), then upgrade using `cluster/local/kind.sh helm-upgrade`. Nothing seems broken. Deletion works as well.
* Installation with the following `values.yaml` to make sure pre-install is not broken:
```yaml
provider:
  packages:
    - crossplane/provider-aws:v0.16.0
    - crossplane/provider-azure:v0.14.0
configuration:
  packages:
    - registry.upbound.io/upbound/platform-ref-gcp:v0.0.1
```



[contribution process]: https://git.io/fj2m9
